monasca_rsyslog/client.py

```diff
@@ -77,20 +77,34 @@ def __init__(self):
         self._min_poll_delay = cfg.CONF.api.min_poll_delay
         self._max_batch_size = cfg.CONF.api.max_batch_size
 
-    def _get_monasca_log_api_url(self, version='v3.0'):
+    def _get_monasca_log_api_url(self):
         """ Retrieves monasca log api endpoint. """
 
-        endpoint = cfg.CONF.api.url
-        if not endpoint:
+        url = cfg.CONF.api.url
+        if not url:
             catalog = self._sess.auth.get_auth_ref(self._sess).service_catalog
             endpoint = catalog.url_for(
                 service_type=cfg.CONF.auth.service_type,
                 interface=cfg.CONF.auth.endpoint_type,
                 region_name=cfg.CONF.auth.region_name,
             )
-        if not endpoint.endswith(version):
-            endpoint += '/' + version
-        return endpoint
+            response = self._sess.get(endpoint)
```

Comment:
+1 to at least put a TODO here saying that this should really use an endpoint filter. We've probably re-implemented part of the endpoint filter below, but without the unit tests and the consideration that the version response can technically yield the information in different formats.

Reply:
Hmm, not sure this is the intended use case for endpoint filter actually because I've just been testing it out and I can for example do things like this:

```
In [91]: c._sess.get('/',endpoint_filter=dict(service_type='logging',interface='public', region_name='RegionOne')).text

Out[91]: u'{"elements":[{"id":"v3.0","links":[{"href":"http:\\/\\/10.60.253.1:5607\\/v3.0\\/logs","rel":"logs"}],"status":"CURRENT","updated":"2016-03-01T00:00:00Z"},{"id":"v2.0","links":[{"href":"http:\\/\\/10.60.253.1:5607\\/v2.0\\/log\\/single","rel":"logs"}],"status":"DEPRECATED","updated":"2015-09-01T00:00:00Z"}],"links":[{"href":"http:\\/\\/10.60.253.1:5607\\/","rel":"self"},{"href":"http:\\/\\/10.60.253.1:5607\\/version","rel":"version"},{"href":"http:\\/\\/10.60.253.1:5607\\/healthcheck","rel":"healthcheck"}]}'
```

So this is equivalent to getting the session query to automatically determine the endpoint URL rather than filtering the results that come back if you see what I mean?

Reply:
I think it should work like this: https://docs.openstack.org/keystoneauth/latest/using-sessions.html#service-discovery

In particular:

```In general a client does not need to know the full URL for the server that they are communicating with, simply that it should send a request to a path belonging to the correct service.

This is controlled by the endpoint_filter parameter to a request which contains all the information an authentication plugin requires to determine the correct URL to which to send a request. When using this mode only the path for the request needs to be specified:
```

So the keystone catalogue has the versionless endpoint, and then we say we want a minimum version of 3 and get it via the endpoint filter.